BlockServer/core/active_config_holder.py

```diff
@@ -216,33 +225,52 @@ def iocs_changed(self):
-        iocs_to_stop = set()
-
-        # Look for any new/changed components
-        for name, component in self._components.items():
-            if name in self._cached_components:
-                _start, _restart = _compare_ioc_properties(
-                    self._cached_components[name], self._components[name])
-
-                iocs_to_start |= _start
-                iocs_to_restart |= _restart
-            else:
-                for ioc_name in component.iocs.keys():
-                    iocs_to_start.add(ioc_name)
-
-        # Look for removed IOCs
-        for ioc_name in set(self._cached_config.iocs.keys()).difference(self._config.iocs.keys()):
-            iocs_to_stop.add(ioc_name)
-
-        # Look for any removed components
-        for name, component in self._cached_components.items():
-            if name not in self._components:
-                for ioc_name in component.iocs.keys():
-                    iocs_to_stop.add(ioc_name)
-
-        return iocs_to_start, iocs_to_restart, iocs_to_stop
+        def _get_config_iocs(config, components):
+            iocs = {}
+            for ioc_name, ioc in config.iocs.items():
+                iocs[ioc_name] = ioc
+
+            for name, component in components.items():
+                for ioc_name, ioc in component.iocs.items():
+                    iocs[ioc_name] = ioc
+            return iocs
+
+        iocs_in_current_config = _get_config_iocs(self._cached_config, self._cached_components)
+        iocs_in_new_config = _get_config_iocs(self._config, self._components)
+
+        new_iocs, changed_iocs, removed_iocs = _compare_ioc_properties(old=iocs_in_current_config,
+                                                                       new=iocs_in_new_config)
+
+        # Look for manually-started IOCS, which have been started with unknown macros and therefore should be assumed
+        # to need stopping or restarting on config change.
+        for ioc_name in get_iocs(CONTROL_SYSTEM_PREFIX):
+            # IOCS which shouldn't be stopped.
+            if any(ioc_name.startswith(x) for x in IOCS_NOT_TO_STOP):
+                continue
+
+            # IOCS which have already been considered as they're part of the cached config or components
+            if ioc_name in iocs_in_current_config:
+                continue
+
+            if self._ioc_control.get_ioc_status(ioc_name) == "RUNNING":
+                if ioc_name in iocs_in_new_config:
+                    # If the IOC is in the new config, we need to restart it as the new config may have macros which
+                    # were not used when the IOC was manually started outside the config.
+                    print_and_log(f"Found manually-started IOC {ioc_name}. Restarting as present in new config.")
+                    if ioc_name in new_iocs:
+                        new_iocs.remove(ioc_name)
+                    changed_iocs.add(ioc_name)
+                else:
```

Comment:
This will relate to a comment in blockserver below. If an ioc is currently running, not in old config, but in new config, then it will get put in "changed" rather than "new" iocs. However when "changed" is only process in  blockserver is autostart is enabled, hence if an ioc is running and is not set to run in the new config it will not get stopped. Adjusting logic to  along the lines of `if {ioc autostart in new} then { add to changed } else { add to stop }` may help, also see comment in blockserver below